internal/probes/main.go

```diff
@@ -0,0 +1,216 @@
+package probes
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/common/slice"
+	"github.com/kuadrant/dns-operator/internal/metrics"
+)
+
+var (
+	ErrInvalidHeader  = fmt.Errorf("invalid header format")
+	ExpectedResponses = []int{200, 201}
+)
+
+const (
+	PROBE_TIMEOUT = 3 * time.Second
+)
+
+type ProbeResult struct {
+	CheckedAt metav1.Time
+	Healthy   bool
+	Reason    string
+	Status    int
+}
+
+func ProcessProbeJob(ctx context.Context, k8sClient client.Client, originalProbe *v1alpha1.DNSHealthCheckProbe) bool {
+	logger := log.FromContext(ctx)
+	probe := &v1alpha1.DNSHealthCheckProbe{}
+	timeUntilProbe := time.Second
+
+	for {
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(originalProbe), probe)
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "error reading probe")
+			return true
+		} else if err != nil {
+			logger.Info("could not find probe CR, killing related worker")
+			return false
+		}
+
+		timeUntilProbe = time.Until(probe.Status.LastCheckedAt.Time.Add(probe.Spec.Interval.Duration))
+		if timeUntilProbe <= 0 {
+			logger.V(1).Info("probe ready to execute", "timeUntilProbe", timeUntilProbe, "lastChecked", probe.Status.LastCheckedAt, "interval", probe.Spec.Interval, "now", time.Now())
+			break
+		}
+
+		logger.V(1).Info("sleeping until probe is ready", "timeUntilProbe", timeUntilProbe, "lastChecked", probe.Status.LastCheckedAt, "interval", probe.Spec.Interval, "now", time.Now())
+		time.Sleep(timeUntilProbe)
+	}
+
+	//if we slept for a while, refresh the probe from the api
+	if timeUntilProbe > time.Second*3 {
```

Comment:
I am not getting this bit, it we sleep it is in the for loop so the probe will get refreshed next time around right?

Reply:
we are not exiting the loop after the sleep so no need to refresh I mean